pkg/package-server/provider/registry.go

```diff
@@ -131,43 +158,176 @@ func (p *RegistryProvider) syncCatalogSource(obj interface{}) (syncError error)
 		return
 	}
 
-	key := sourceKey{source.GetName(), source.GetNamespace()}
-	client, ok := p.getClient(key)
-	if ok && source.Status.RegistryServiceStatus.ServiceName != "" {
-		logger.Info("update detected, attempting to reset grpc connection")
-		client.conn.ResetConnectBackoff()
+	address := source.Address()
+	logger = logger.WithField("address", address)
 
-		ctx, cancel := context.WithTimeout(context.TODO(), defaultConnectionTimeout)
-		defer cancel()
+	key := resolver.CatalogKey{
+		Namespace: source.GetNamespace(),
+		Name:      source.GetName(),
+	}
+	if sourceMeta := p.sources.GetMeta(key); sourceMeta != nil && sourceMeta.Address == address {
+		// If the address hasn't changed, don't bother creating a new source
+		logger.Debug("catalog address unchanged, skipping source creation")
+		return
+	}
 
-		changed := client.conn.WaitForStateChange(ctx, connectivity.TransientFailure)
-		if !changed {
-			logger.Debugf("grpc connection reset timeout")
-			syncError = fmt.Errorf("grpc connection reset timeout")
-			return
+	logger.Info("connecting to source")
+	if _, syncError = p.sources.Add(key, address); syncError != nil {
+		logger.Warn("failed to create a new source")
+	}
+
+	return
+}
+
+func (p *RegistryProvider) syncSourceState(state registrygrpc.SourceState) {
+	key := state.Key
+	logger := logrus.WithFields(logrus.Fields{
+		"action": "sync source",
+		"source": key,
+		"state":  state.State,
+	})
+	logger.Debug("source state changed")
+
+	timeout, cancel := context.WithTimeout(context.Background(), cacheTimeout)
+	defer cancel()
+
+	var err error
+	switch state.State {
+	case connectivity.Ready:
+		var client *registryClient
+		client, err = p.registryClient(key)
+		if err == nil {
+			err = p.refreshCache(timeout, client)
 		}
+	case connectivity.TransientFailure, connectivity.Shutdown:
```

Comment:
I wonder if having transient failure here actually adds additional work. Did you find that gc-ing a source and then re-adding was quicker than letting a source transition back to connecting and (hopefully) ready?

Reply:
We discussed offline that having something in the cache in this state is probably not the best, so we'll just leave it as is.